### PR TITLE
SubArray: Fix scalar indexing with non-Int values

### LINKED ISSUE #15168: Indexing SubArrays with Int32 returns zero dimensional array
I've seen this error for the first time today when building RPM nightlies. It happened twice, both on 64-bit. This is with system LLVM 3.7.1 + Julia patches and ORCJIT.
```
	From worker 3:	     * linalg/cholesky      Error During Test
	From worker 3:	  Test threw an exception of type MethodError
	From worker 3:	  Expression: norm(apd * (cpapd \ b) - b) / norm(b) <= Îµ * Îº * n
	From worker 3:	  MethodError: Cannot `convert` an object of type Array{Float32,0} to an object of type Float32
	From worker 3:	  This may have arisen from a call to the constructor Float32(...),
	From worker 3:	  since type constructors fall back to convert methods.
	From worker 3:	  Closest candidates are:
	From worker 3:	    convert(::Type{Float32}, !Matched::Int8)
	From worker 3:	    convert(::Type{Float32}, !Matched::Int16)
	From worker 3:	    convert(::Type{Float32}, !Matched::Int32)
	From worker 3:	    ...
	From worker 3:	   [inlined code] from ./range.jl:419
	From worker 3:	   in setindex!(::SubArray{Float32,1,Array{Float32,2},Tuple{UnitRange{Int64},Int64},true}, ::Array{Float32,0}, ::Int64) at ./subarray.jl:182
	From worker 3:	   [inlined code] from ./abstractarray.jl:341
	From worker 3:	   in permute!!(::SubArray{Float32,1,Array{Float32,2},Tuple{UnitRange{Int64},Int64},true}, ::Array{Int32,1}) at ./combinatorics.jl:82
	From worker 3:	   [inlined code] from ./array.jl:64
	From worker 3:	   in A_ldiv_B!(::Base.LinAlg.CholeskyPivoted{Float32,Array{Float32,2}}, ::Array{Float32,2}) at ./linalg/cholesky.jl:252
	From worker 3:	   in \(::Base.LinAlg.CholeskyPivoted{Float32,Union{DenseArray{Float32,2},SubArray{Float32,2,A<:DenseArray{T,N},I<:Tuple{Vararg{Union{Base.AbstractCartesianIndex{N},Base.NoSlice,Colon,Int64,Range{Int64}}}},L}}}, ::Array{Float32,2}) at ./linalg/factorization.jl:39
	From worker 3:	   [inlined code] from /builddir/build/BUILD/julia/test/linalg/cholesky.jl:116
	From worker 3:	   in anonymous at ./no file:4294967295
	From worker 3:	   [inlined code] from ./essentials.jl:78
	From worker 3:	   in include_string(::UTF8String, ::ASCIIString) at ./loading.jl:371
	From worker 3:	   in include_from_node1(::ASCIIString) at ./loading.jl:420
	From worker 3:	   [inlined code] from ./util.jl:179
	From worker 3:	   in runtests(::ASCIIString) at /builddir/build/BUILD/julia/test/testdefs.jl:7
	From worker 3:	   in (::Base.Serializer.__deserialized_types__.##9046)(::ASCIIString) at /builddir/build/BUILD/julia/test/runtests.jl:36
	From worker 3:	   in run_work_thunk(::Base.##254#256{Base.CallMsg{:call_fetch}}, ::Bool) at ./multi.jl:714
	From worker 3:	   [inlined code] from ./multi.jl:1010
	From worker 3:	   in (::Base.##253#255{Base.CallMsg{:call_fetch},TCPSocket})() at ./task.jl:59
Exception running test linalg/cholesky :
On worker 3:
LoadError: There was an error during testing
 in record at ./test.jl:317
 in do_test at ./test.jl:215
 [inlined code] from ./test.jl:137
 in anonymous at ./no file:4294967295
 [inlined code] from ./essentials.jl:78
 in include_string at ./loading.jl:371
 in include_from_node1 at ./loading.jl:420
 [inlined code] from ./util.jl:179
 in runtests at /builddir/build/BUILD/julia/test/testdefs.jl:7
 in #16 at /builddir/build/BUILD/julia/test/runtests.jl:36
 in run_work_thunk at ./multi.jl:714
 [inlined code] from ./multi.jl:1010
 in #253 at ./task.jl:59
while loading /builddir/build/BUILD/julia/test/linalg/cholesky.jl, in expression starting on line 24ERROR: LoadError: Some tests exited with errors.
 in (::##11#19)() at /builddir/build/BUILD/julia/test/runtests.jl:64
 in cd(::##11#19, ::ASCIIString) at ./file.jl:47
 in include(::ASCIIString) at ./boot.jl:264
 in include_from_node1(::UTF8String) at ./loading.jl:417
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
while loading /builddir/build/BUILD/julia/test/runtests.jl, in expression starting on line 13
WARNING: Forcibly interrupting busy workers
ArgumentError("stream is closed or unusable")
```
https://copr-be.cloud.fedoraproject.org/results/nalimilan/julia-nightlies/epel-7-x86_64/00161544-julia/build.log.gz
https://copr-be.cloud.fedoraproject.org/results/nalimilan/julia-nightlies/fedora-rawhide-x86_64/00161544-julia/build.log.gz

### PR DESCRIPTION
**Pull request by mbauman**
Fix #15168.

~~This is a quick patch — it still needs tests and to actually run the tests.~~

### CODE REVIEW

File: base/subarray.jl [hunk 1]
```diff
@@ -143,43 +143,43 @@ reindex(V, idxs::Tuple{DroppedScalar, Any, Vararg{Any}}, subidxs::Tuple{Any, Any
 
 # In general, we simply re-index the parent indices by the provided ones
 getindex(V::SubArray) = (@_propagate_inbounds_meta; getindex(V, 1))
-function getindex(V::SubArray, I::Int...)
+function getindex(V::SubArray, I::Real...)
```
Comment (andreasnoack) [inline base/subarray.jl]: Didn't we deprecate non-integer indexing?

Comment (mbauman) [inline base/subarray.jl]: We did; the deprecation is still in effect, though, so this way SubArray behaves as it did before my refactor and still issues warnings.

**Comment by nalimilan** [comment]
Looks like it fixed the bug in the cholesky test, but I get many failures related to `to_index`:
```
	From worker 5:	  Test threw an exception of type ErrorException
	From worker 5:	  Expression: scale([1,2],sub(a,:,:)) == a .* [1,2]
	From worker 5:	  to_index(I::Tuple) is deprecated, use to_indexes(I...) instead.
	From worker 5:	   in depwarn(::ASCIIString, ::Symbol) at ./deprecated.jl:70
	From worker 5:	   in to_index(::Tuple{Int64,Int64}) at ./deprecated.jl:448
	From worker 5:	   [inlined code] from ./subarray.jl:173
	From worker 5:	   in scale!(::Array{Float64,2}, ::Array{Int64,1}, ::SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Colon},true}) at ./linalg/matmul.jl:37
	From worker 5:	   in scale(::Array{Int64,1}, ::SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Colon},true}) at ./linalg/matmul.jl:42
	From worker 5:	   [inlined code] from /builddir/build/BUILD/julia/test/linalg/generic.jl:114
	From worker 5:	   in anonymous at ./no file:4294967295
	From worker 5:	   [inlined code] from ./essentials.jl:78
	From worker 5:	   in include_string(::UTF8String, ::ASCIIString) at ./loading.jl:371
	From worker 5:	   in include_from_node1(::ASCIIString) at ./loading.jl:420
	From worker 5:	   [inlined code] from ./util.jl:179
	From worker 5:	   in runtests(::ASCIIString) at /builddir/build/BUILD/julia/test/testdefs.jl:7
	From worker 5:	   in (::Base.Serializer.__deserialized_types__.##9181)(::ASCIIString) at /builddir/build/BUILD/julia/test/runtests.jl:36
	From worker 5:	   in run_work_thunk(::Base.##254#256{Base.CallMsg{:call_fetch}}, ::Bool) at ./multi.jl:714
	From worker 5:	   [inlined code] from ./multi.jl:1010
	From worker 5:	   in (::Base.##253#255{Base.CallMsg{:call_fetch},TCPSocket})() at ./task.jl:59
```
See the other ones at https://copr-be.cloud.fedoraproject.org/results/nalimilan/julia-nightlies/fedora-rawhide-x86_64/00162509-julia/build.log.gz

**Comment by nalimilan** [comment]
What's new is that it also broke i386, which was working fine. But it's weird that deprecations look like they are considered as errors...

**Comment by mbauman** [comment]
Not terribly weird.  I made this change blindly — without compiling or testing myself.  This is the right answer, I just missed a splat or something.  I'll fix it properly tonight.

**Comment by nalimilan** [comment]
Great, thanks!

**Comment by mbauman** [comment]
Alright, updated and fixed.

**Comment by nalimilan** [comment]
Thanks, the new fix passes the tests on all platforms.

**Pull request merged**